src/Microsoft.Identity.Web.TokenAcquisition/ConfidentialClientApplicationBuilderExtension.cs

```diff
@@ -30,23 +30,61 @@ public static async Task<ConfidentialClientApplicationBuilder> WithClientCredent
             IEnumerable<CredentialDescription> clientCredentials,
             ILogger logger,
             ICredentialsLoader credentialsLoader,
-            CredentialSourceLoaderParameters credentialSourceLoaderParameters)
+            CredentialSourceLoaderParameters? credentialSourceLoaderParameters)
+        {
+            var credential = await LoadCredentialForMsalOrFailAsync(
+                    clientCredentials,
+                    logger,
+                    credentialsLoader,
+                    credentialSourceLoaderParameters)
+                .ConfigureAwait(false);
+
+            if (credential == null)
+            {
+                return builder;
+            }
+
+            switch (credential.CredentialType)
+            {
+                case CredentialType.SignedAssertion:
+                    return builder.WithClientAssertion((credential.CachedValue as ClientAssertionProviderBase)!.GetSignedAssertionAsync);
+                case CredentialType.Certificate:
+                    return builder.WithCertificate(credential.Certificate);
+                case CredentialType.Secret:
+                    return builder.WithClientSecret(credential.ClientSecret);
+                default:
+                    throw new NotImplementedException();
+
+            }
+        }
+
+        internal /* for test */ async static Task<CredentialDescription?> LoadCredentialForMsalOrFailAsync(
```

Comment:
Logic remains mostly the same, I just separated the existing method into 2, so that I can unit test.